### PR TITLE
feat: add `Type::as_struct`

### PR DESCRIPTION
**Pull request by asterite**
# Description

## Problem

Part of #5668

## Additional Context

None.

## Documentation

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: test_programs/compile_success_empty/comptime_type/src/main.nr [hunk 1]
```diff
@@ -66,5 +70,17 @@ fn main() {
...
         assert(bool_type.is_bool());
+
+        // Check Type::as_struct
+        assert(u8_type.as_struct().is_none());
+
+        let foo = Foo { x: 0 };
+        let foo_type = type_of(foo);
+        let (struct_definition, generics) = foo_type.as_struct().unwrap();
+        let fields = struct_definition.fields();
+        assert_eq(fields.len(), 1);
```
Comment (asterite) [inline test_programs/compile_success_empty/comptime_type/src/main.nr:81]: If I check `fields[0].1`, which is the type of the field, I get it's `T`. I wanted to assert that's it's the field type. Is the current behavior okay? I didn't know if `as_struct()` would return the uninstantiated struct type, or the instantiated one (it currently returns the uninstantiated one).

File: compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs [hunk 2]
```diff
@@ -550,6 +551,27 @@ fn type_as_slice(
...
+    return_type: Type,
+    location: Location,
+) -> IResult<Value> {
+    type_as(arguments, return_type, location, |typ| {
+        if let Type::Struct(struct_type, generics) = typ {
+            Some(Value::Tuple(vec![
+                Value::StructDefinition(struct_type.borrow().id),
+                Value::Slice(
+                    generics.into_iter().map(Value::Type).collect(),
+                    Type::Quoted(QuotedType::Type),
```
Comment (jfecher) [inline compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs:566]: This should be the slice type, not the element type

File: test_programs/compile_success_empty/comptime_type/src/main.nr (hunk 1 above)
Comment (jfecher) [reply test_programs/compile_success_empty/comptime_type/src/main.nr:81]: The only way to get the instantiated one is to use the generics to re-instantiate it. Ie. wrap it in a `Type::Struct` again.
We could add a method `get_struct_fields(self) -> Option<[(Quoted, Type)]>` on `Type` to get the instantiated struct field's.

**Review by jfecher — CHANGES_REQUESTED** [review]
Need to update the slice type but otherwise LGTM

File: compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs (hunk 2 above)
Comment (asterite) [reply compiler/noirc_frontend/src/hir/comptime/interpreter/builtin.rs]: Done!

**Review by jfecher — APPROVED** [review]
:+1:

**Pull request merged**